### ISSUE TITLE
Roadmap

### ISSUE THREAD

**Issue by Gigas002** [Gigas002/AntiDupl#1]
Recently I've started working on some features and wrote a roadmap for updating the repo with some plans for future. After listening to the opinions of the author of AntiDupl (@ermig1979 ) and the community, I'll contribute to this repo or I'll continue on developing the app in my fork, if you don't think, that proposed changes will affect good on this project.

**Main points:**

- [x] Drop support of `x86` and end-of-life OS (*these users can still use outdated versions anyway*);
- [x] Move to latest stable .NET for dotnet projects (*since we can publish portable, self-contained apps this change shouldn't matter for users in case of pre-installed frameworks*);
- [x] Support for [SemVer](https://semver.org);
- [ ] Add CI/CD for dotnet projects in the future (*GitHub CI or Appveyor*);
- [ ] Clean up repo (*close outdated/resolved issues, write issue/pr-templates, add milestones, projects, write a proper CHANGELOG.md, etc*);

**Core:**

The purpose of changing the core is to separate logics in solition. We shouldn't build c++ project with all of it's dependencies for each time we want to debug program. And we also don't need to have heavy c++ developer's tools if we're working on dotnet GUI only. The core itself should be reusable and have one code base for all the other projects. So, the points are following:

- [x] Move `AntiDupl` dependencies out of project to simplify build and the process of updating dependencies (*I suggest doing it through [vcpkg](https://docs.microsoft.com/ru-ru/cpp/build/vcpkg)*):
- [x] libwebp (1.1.0#1);
- [x] libopenjpeg (2.3.1#2);
- [x] turbojpeg (2.0.5#1);
- [x] prebuilt simd;
- [ ] vcpkg simd (*[waiting](https://github.com/microsoft/vcpkg/issues/15008) for it to be added to vcpkg packages feed*);
- [x] Create nuget package with native libraries (*AntiDupl.Native.Win-x64*);
- [x] Move bindings with `unsafe` code from current `AntiDupl.NET` to new `AntiDupl.NET.Core` library;
- [ ] Build and publish `AntiDupl.NET.Core` separately as new nuget package (*so it can be reusable in other apps without rebuilding and with one code base*);
- [ ] Write unit tests for `AntiDupl.NET.Core` (*or AntiDupl c++ library itself; needs help!*);

Unfortunately, I don't have a lot of experience in writing c++ code, so I can't help with fixing core bugs/improving core performance/etc. Because of this, for the part with tests we'll definitely need some help. I suggest writing tests for bindings (`AntiDupl.NET.Core`, not `AntiDupl`) on C# to not complicate CI/CD in the future and make these operations with dotnet/pwsh-scripting only.

**WinForms app:**

- [x] Move `AntiDupl.NET` to `AntiDupl.NET.WinForms` namespace and rename the project itself;
- [x] Remove unsafe code and use `AntiDupl.NET.Core` with `AntiDupl.Native` packages as dependencies;
- [ ] Update and beautify all dotnet code in project (*naming conventions, moving Strings.cs to Strings.resx with Strings.ru-ru.resx and others as localization, etc*);
- [ ] Release updated app as 2.3.11 or 2.4.0 (*last release of WinForms app?*);

The point of these changes is to simplify protracted transition to **WPF** and develop only it in future. From my perspective, for first new major `3.0.0` release we should rewrite WinForms app as WPF straightforward as it is, and then we can move to a better architecture, like MVVM.

**Comment by Gigas002** [comment]
Talking about x86 versions of projects, at current state it's easily to add them back again (only configuration changes needed, no adding new files, etc), but I just don't see the point.

**Comment by ermig1979** [comment]
I am glad to see activities at AntiDupl project. The many items above are usefull and I was going to do some of them. In particular to remain only x64 support, replace main functionality form C++ to C#. I hope it will not only plans.

**Comment by Gigas002** [comment]
Thank you for your answer! I'm in process of implementation these features, yet I can't tell yet, when I'll be able to create a solid PR. For now I hope to complete most part until the end of 2021's January.

**Comment by Gigas002** [comment]
Some progress reports here.

TLDR, nearest goals for now:
- Publish native nuget package (almost done);
- Upgrade cpp bindings logics into dotnet core bindings library (in progress);
- Upgrade the way options are handled in WinForms app;
- Publish bindings package to nuget;
- Configure ci autobuilds;

All the other stuff goes on later, that's the first step.

Currently I'm working on upgrading binding's core logics. I've encountered some difficulties and also lack time, so the progress goes on very slow, but I'm not abandoning the project. I've decided to rewrite the way, `Ad...Options` handled in application, so in my dev branch it's functionality partly broken atm (`master` branch is still safe to use for now).
Packaging cpp core into native nuget package going on pretty well, I'm ready to publish it to nuget already. I'd like to use vcpkg's `simd` before it, but no one seems to grab on my issue on adding `simd` into vcpkg package feed (I don't know how to do it myself unfortunately).
I also thought about WPF app, and from my perspective it would be better to write Avalonia GUI instead of WPF. It's THEORETICALLY possible to port the cpp core to linux, so writing cross-platofrm gui from the start would result into less pain in the future.

**Comment by Edi61** [comment]
Hi Gigas002,

are you still working on this? I have implemented the support of simd in vcpkg, so its now available. It should directly work, after removing all simd stuff from AntiDupl.
https://github.com/microsoft/vcpkg/pull/28163

**Comment by Gigas002** [comment]
@Edi61 Sorry for a late response. Unfortunately, as you can see, I've stopped working on project due to some personal difficulties (plus, I've moved to linux on my personal pc, so I can't even build winforms/wpf projects without VMs). Feel free to contact me if you have any questions about the stages I have already done. Some stuff, like moving to `AntiDupl.Native` package is easy to implement, but I haven't completed the bindings for `AntiDupl.NET.Core` package

**Comment by Edi61** [comment]
Hi @Gigas002,
thanks for response

**Comment by Gigas002** [comment]
@Edi61 Great to see a lot of work being done and my fork was helpful at least for a bit 👍! 
Since this a discussable question, I wasn't sure where to put it, so decided not to create a new issue.
IMO it would be a good architecture change to publish `AntiDupl.Native` and `AntiDupl.NET.Core` packages onto nuget. That would simplify the building process for people, who doesn't want to build c++ portion of the code, install c++ toolset and just want to work on GUI apps. Plus, that could simplify the writing of CI/CD. What do you think about that?

**Comment by Edi61** [comment]
Hello @Gigas002,
I am not jet familiar with Nuget so much, but your idea sounds plausible. At least I have also seen in the discussions that they seem to be interested people just on the AntiDupl.Native.
Could you describe a little more how a nuget publish procedure would look like and also how your concrete idea was regarding CI/CD?

**Comment by Gigas002** [comment]
@Edi61 Native package, in fact, is doesn't contain any code with exception of `.csproj` file. The example for `AntiDupl.Native` in [my repo](https://github.com/Gigas002/AntiDupl/blob/master/src/AntiDupl.Native/AntiDupl.Native.csproj) should work just fine. 
First `PropertyGroup` just defines package's metadata.
Next block controls `Release`-specific metadata. The nuget package will automatically be built on release if `GeneratePackageOnBuild` property is set to `true`.
The last `ItemGroup` block sets a path to a binaries to put into library (pre-built c++ `AntiDupl` binaries), e.g. `<None Include="../AntiDupl/bin/$(Platform)/Release/*.dll">`. After that, the package must be manually put onto https://www.nuget.org/. It's possible to auto-update it with CI/CD afterwards, but the first push must be manual, I think.
I also think, that `AntiDupl` (C++ project) and `AntiDupl.Native` should have it's own configurations (e.g. `<Configurations>Native-Debug;Native-Release</Configurations>`, so we can separate them from build in C#-only CI/CD config.
I'm not familiar with CI/CD for C++ projects, but as we separate native part from C# code I think I'll contribute to a repo. If you want to implement it by yourself, it's pretty easy to do with GitHub's CI/CD tools. Since this project doesn't have any tests, the `yml` code would look something like this:

```yml
name: .NET

on: [push, pull_request]

jobs:

# Build on Windows

  build-windows:

    runs-on: windows-latest
    
    steps:
    - uses: actions/checkout@v3.0.2
    - name: Setup .NET
      uses: actions/setup-dotnet@v2.1.0
      with:
        dotnet-version: 6.0.x
    - name: Restore dependencies
      run: dotnet restore
    - name: Build
      run: dotnet build --no-restore -c Release /p:Platform=x64
```

for a build tasks. Now, for CD, I haven't written this with GitHub (though I don't think would be hard to implement), but I've used AppVeyor. It uses `yml` files as config as well, it's syntax is pretty straightforward and it also has a web GUI for creating config.
